src/harness/sas_objects.py

```diff
@@ -0,0 +1,353 @@
+
+    def cbsd_id_has_atleast_one_grant(self):
+        """
+
+        :rtype: bool
+        """
+        for grant_id, grant_request in self.grant_objects.items():
+            if grant_request.get_grant_authorized_in_last_heartbeat():
+                return True
+        return False
+
+    def construct_heartbeat_request_for_all_active_grant(self):
+        """
+
+        :rtype: heartbeat_requests
+        """
+        heartbeat_requests = []
+        for grant_id, grant_object in self.grant_objects.items():
+            heartbeat_request = grant_object.construct_heartbeat_request()
+            heartbeat_requests.append(heartbeat_request)
+        return heartbeat_requests
+
+    def get_all_operation_params_of_grant_request(self):
+        """
+
+        :rtype: operation_params
+        """
+        operation_params = []
+        for grant_id, grant_object in self.grant_objects.items():
+            if True == grant_object.get_grant_authorized_in_last_heartbeat():
+                operation_param = grant_object.get_request_operation_param()
+                operation_params.append(operation_param)
+        return operation_params
+
+
+class DomainProxy:
+
+    """
+    A DP object has the following methods
+        __init__
+                Parameters: certificate, N registration requests, N grant requests
+                Send N registration requests, expect success for all
+                Send N grant requests, allow failure
+
+        Retrieve CBSD object with CBSD ID = C
+        Heartbeat all CBSDs/Grants (and update internal state)
+        Heartbeat/relinquish/grant/heartbeat request for all CBSDs (and update internal state)
+            Heartbeat on all active grants (allow all failures)
+            If any HBs contain suggested operating params, relinquish the corresponding grant
+            For all grants in the previous step, request a new grant with the suggested operating parameters
+                            (allow failure)
+            Heartbeat all active grants (allow all failures)
+        Retrieve all CBSD objects with at least one grant which is authorized to transmit
+    """
+
+    def __init__(self, certificate, key, registration_requests, grant_requests):
+        # checking for the number of registration requests matches number of grant requests
+        if len(registration_requests) != len(grant_requests):
+            return
+        self.dp_certificate = certificate
+        self.dp_key = key
+        self.cbsd_objects = {}
+        cbsd_id = []
+        self.sas_admin = sas.GetTestingSas()
+        for registerRequest in registration_requests[:]:
+            response = self.sas_admin.Registration(registerRequest, self.dp_certificate, self.dp_key)\
+                                                                  ['registrationResponse'][0]
+            cbsd_id.append(response['cbsdId'])
+            self.sas_admin.assertEqual(response['response']['responseCode'], 0)
```

Comment:
This makes no sense whatsoever -- the function `assertEqual` is part of the testcase framework and unittest class, and does not belong to the SAS Admin interface.

I'm going to stop reviewing at this point until the code is internally tested.

Reply:
This is   corrected. Thanks for pointing it out. We have changed the input parameter to   the invoking sas_testcase object and used that object for assert methods.